packages/jest-haste-map/src/index.js

```diff
@@ -0,0 +1,370 @@
+          `${existingModule[H.PATH]}\n\n` +
+          `This warning is caused by a @providesModule declaration ` +
+          `with the same name accross two different files.`
+        );
+        return;
+      }
+
+      moduleMap[platform] = module;
+    };
+
+    for (const filePath in hasteMap.files) {
+      if (mocksPattern && mocksPattern.test(filePath)) {
+        mocks[path.basename(filePath, path.extname(filePath))] = filePath;
+      }
+
+      const fileMetadata = hasteMap.files[filePath];
+      const moduleMetadata = hasteMap.map[fileMetadata[H.ID]];
+      if (fileMetadata[H.VISITED]) {
+        if (!fileMetadata[H.ID]) {
+          continue;
+        } else if (fileMetadata[H.ID] && moduleMetadata) {
+          map[fileMetadata[H.ID]] = moduleMetadata;
+          continue;
+        }
+      }
+
+      promises.push(
+        this._getWorker()({filePath}).then(
+          metadata => {
+            fileMetadata[H.VISITED] = 1;
+            if (metadata.id) {
+              fileMetadata[H.ID] = metadata.id;
+              setModule(metadata.id, metadata.module);
+            }
+            fileMetadata[H.DEPENDENCIES] = metadata.dependencies || [];
+          },
+          error => {
+            // If a file cannot be read we remove it from the file list and
+            // ignore the failure silently.
+            delete hasteMap.files[filePath];
+          }
+        )
+      );
+    }
+
+    return Promise.all(promises)
+      .then(() => {
+        if (this._workerFarm) {
+          workerFarm.end(this._workerFarm);
+        }
+        this._workerFarm = null;
+        this._workerPromise = null;
+      })
+      .then(() => {
+        hasteMap.map = map;
+        hasteMap.mocks = mocks;
+        return hasteMap;
+      });
+  }
+
+  /**
+   * 4. serialize the new `HasteMap` in a cache file.
+   */
+  _persist(hasteMap) {
+    fs.writeFileSync(this._cachePath, JSON.stringify(hasteMap), 'utf-8');
+    return hasteMap;
+  }
+
+  /**
+   * Creates workers or parses files and extracts metadata in-process.
+   */
+  _getWorker() {
+    if (!this._workerPromise) {
+      if (this._options.maxWorkers === 1) {
+        this._workerPromise = message => new Promise(
+          (resolve, reject) => worker(message, (error, metadata) => {
+            if (error) {
+              reject(error);
+            } else {
+              resolve(metadata);
+            }
+          })
+        );
+      } else {
+        this._workerFarm = workerFarm(
+          {
+            maxConcurrentWorkers: this._options.maxWorkers,
+          },
+          require.resolve('./worker')
+        );
+        this._workerPromise = denodeify(this._workerFarm);
+      }
+    }
+
+    return this._workerPromise;
+  }
+
+  _parse(hasteMap) {
+    hasteMap = JSON.parse(hasteMap);
+    for (const key in hasteMap) {
+      Object.setPrototypeOf(hasteMap[key], null);
+    }
+    return hasteMap;
+  }
+
+  _crawl(hasteMap) {
```

Comment:
Are you going to make buck a dependency of packager even for open source? In either case, I think this should be pretty easy to support: the options passed into an instance can support a `crawl` function and for buck supported packager it can be as simple as `crawl: () => Promise.resolve(filesFromBuck);`.